devito/ir/equations/algorithms.py

```diff
@@ -162,74 +163,123 @@ def concretize_subdims(exprs, **kwargs):
     """
     sregistry = kwargs.get('sregistry')
 
-    # {root Dimension -> {SubDimension -> concrete SubDimension}}
-    mapper = defaultdict(dict)
+    mapper = {}
+    rebuilt = {}  # Rebuilt implicit dims etc which are shared between dimensions
 
-    _concretize_subdims(exprs, mapper, sregistry)
+    _concretize_subdims(exprs, mapper, rebuilt, sregistry)
     if not mapper:
         return exprs
 
-    subs = {}
-    for i in mapper.values():
-        subs.update(i)
+    # There may be indexed Arrays defined on SubDimensions in the expressions
+    # These must have their dimensions replaced and their .function attribute
+    # reset to prevent recovery of the original SubDimensions
+    functions = {f for f in retrieve_functions(exprs) if f.is_Array}
+    for f in functions:
+        dimensions = tuple(mapper[d] if d in mapper else d for d in f.dimensions)
+        if dimensions != f.dimensions:
+            # A dimension has been rebuilt, so build a mapper for Indexed
+            mapper[f.indexed] = f._rebuild(dimensions=dimensions, function=None).indexed
```

Comment:
Shouldn't `f` already be indexed if getting there?

Reply:
No, `retrieve_functions` gets the underlying function

Reply:
then shouldn't this use `retrieve_indexed` instead?

Reply:
No, because you can `uxreplace` a single `{f.indexed: g.indexed}` to turn all `f[x]`, `f[x+1]` into `g[x]`, `g[x+1]`